File: DSAlgorithms/Leetcode/c++/sliding_window/239-sliding_window_max.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
       std::vector<int> result; 
       std::deque<int> dq;  //contains indices of nums
       int left = 0;
       int right = 0;
       while(right < nums.size())
       {
           while(!dq.empty() && nums[dq.back()] < nums[right])
                dq.pop_back();
            dq.push_back(right);
            
            if(left > dq.front())
                dq.pop_front();

            if((right + 1) >= k)
            {
                result.push_back(nums[dq.front()]);
                left ++;
            }
            right++;
       }
       return result;
    }
};

File: dsa_algo/leetcode/ts/src/sliding_window/239-sliding_window_max.ts
const maxSlidingWindow = (nums: number[], k: number): number[] =>{
    //eh, js doesn't have a built in deque. using arr for now
    //it works so whatever
    const result : number[] = [];
    const dq: number[] = [];
    let left = 0;
    let right = 0;
    while(right < nums.length)
    {
        while((dq.length > 0) && nums[dq[dq.length - 1]] < nums[right])
            dq.pop();
        dq.push(right);
        if(left > dq[0])
            dq.shift();
        if((right + 1) >= k)
        {
            result.push(nums[dq[0]])
            left++;
        }
        right++;
    }
    return result;
};